ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/storage/keyvalue/RocksDbStorageProvider.java

```diff
@@ -12,22 +12,83 @@
+
 import tech.pegasys.pantheon.ethereum.storage.StorageProvider;
 import tech.pegasys.pantheon.metrics.MetricsSystem;
+import tech.pegasys.pantheon.services.kvstore.ColumnarRocksDbKeyValueStorage;
 import tech.pegasys.pantheon.services.kvstore.KeyValueStorage;
 import tech.pegasys.pantheon.services.kvstore.RocksDbConfiguration;
 import tech.pegasys.pantheon.services.kvstore.RocksDbKeyValueStorage;
+import tech.pegasys.pantheon.services.kvstore.SegmentedKeyValueStorage;
+import tech.pegasys.pantheon.services.kvstore.SegmentedKeyValueStorage.Segment;
+import tech.pegasys.pantheon.services.kvstore.SegmentedKeyValueStorageAdapter;
 
 import java.io.IOException;
 import java.nio.file.Files;
 
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
 public class RocksDbStorageProvider {
+  private static final Logger LOG = LogManager.getLogger();
 
   public static StorageProvider create(
       final RocksDbConfiguration rocksDbConfiguration, final MetricsSystem metricsSystem)
       throws IOException {
+    if (rocksDbConfiguration.useColumns()) {
+      return createSegmentedProvider(rocksDbConfiguration, metricsSystem);
+    } else {
+      return createUnsegmentedProvider(rocksDbConfiguration, metricsSystem);
+    }
+  }
+
+  private static StorageProvider createUnsegmentedProvider(
+      final RocksDbConfiguration rocksDbConfiguration, final MetricsSystem metricsSystem)
+      throws IOException {
     Files.createDirectories(rocksDbConfiguration.getDatabaseDir());
     final KeyValueStorage kv = RocksDbKeyValueStorage.create(rocksDbConfiguration, metricsSystem);
     return new KeyValueStorageProvider(kv);
   }
+
+  private static StorageProvider createSegmentedProvider(
+      final RocksDbConfiguration rocksDbConfiguration, final MetricsSystem metricsSystem)
+      throws IOException {
+    LOG.info("Using RocksDB columns");
+    Files.createDirectories(rocksDbConfiguration.getDatabaseDir());
+    final SegmentedKeyValueStorage<?> columnarStorage =
+        ColumnarRocksDbKeyValueStorage.create(
+            rocksDbConfiguration, asList(RocksDbSegment.values()), metricsSystem);
+
+    return new KeyValueStorageProvider(
+        new SegmentedKeyValueStorageAdapter<>(RocksDbSegment.BLOCKCHAIN, columnarStorage),
+        new SegmentedKeyValueStorageAdapter<>(RocksDbSegment.WORLD_STATE, columnarStorage),
+        new SegmentedKeyValueStorageAdapter<>(RocksDbSegment.PRIVATE_TRANSACTIONS, columnarStorage),
+        new SegmentedKeyValueStorageAdapter<>(RocksDbSegment.PRIVATE_STATE, columnarStorage),
+        new SegmentedKeyValueStorageAdapter<>(RocksDbSegment.PRUNING_STATE, columnarStorage));
+  }
+
+  private enum RocksDbSegment implements Segment {
+    BLOCKCHAIN(new byte[] {1}),
+    WORLD_STATE(new byte[] {2}),
+    PRIVATE_TRANSACTIONS(new byte[] {3}),
+    PRIVATE_STATE(new byte[] {4}),
+    PRUNING_STATE(new byte[] {5});
+
+    private final byte[] id;
+
+    RocksDbSegment(final byte[] id) {
```

Comment:
nit: would be tidier if we made this accept `final byte... id` then the array would be created automatically and the enum declarations would just be `BLOCKCHAIN(1), WORLD_STATE(2)` etc.  Cheating but kind of nice. :)